File: Talleres RESFTUL JAVA DOTNET/API_GOOGLE_MAPS_REST_JAVA_GR-03/04. SERVIDOR/api-java-bank/src/main/java/com/allpasoft/apijavabank/web/controller/AccountController.java
package com.allpasoft.apijavabank.web.controller;

import com.allpasoft.apijavabank.application.dto.AccountDto;
import com.allpasoft.apijavabank.application.dto.CreateAccountDto;
import com.allpasoft.apijavabank.application.dto.UpdateAccountDto;
import com.allpasoft.apijavabank.application.service.AccountService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/account")
@RequiredArgsConstructor
@Tag(name = "Cuentas", description = "Endpoints para gestión de cuentas bancarias")
public class AccountController {

    private final AccountService accountService;

    @GetMapping
    @Operation(
        summary = "Listar todas las cuentas",
        description = "Obtiene la lista de todas las cuentas bancarias",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de cuentas obtenida exitosamente")
        }
    )
    public ResponseEntity<List<AccountDto>> getAll() {
        return ResponseEntity.ok(accountService.getAll());
    }

    @GetMapping("/{id}")
    @Operation(
        summary = "Obtener cuenta por ID",
        description = "Obtiene una cuenta bancaria por su ID",
        responses = {
            @ApiResponse(responseCode = "200", description = "Cuenta obtenida exitosamente"),
            @ApiResponse(responseCode = "404", description = "Cuenta no encontrada")
        }
    )
    public ResponseEntity<Object> getById(@PathVariable Long id) {
        return accountService.getById(id)
                .<ResponseEntity<Object>>map(ResponseEntity::ok)
                .orElse(ResponseEntity.status(404).body("Cuenta no encontrada"));
    }

    @GetMapping("/user/{userId}")
    @Operation(
        summary = "Obtener cuentas por usuario",
        description = "Obtiene todas las cuentas de un usuario",
        responses = {
            @ApiResponse(responseCode = "200", description = "Cuentas obtenidas exitosamente")
        }
    )
    public ResponseEntity<List<AccountDto>> getByUserId(@PathVariable Long userId) {
        return ResponseEntity.ok(accountService.getByUserId(userId));
    }

    @PostMapping
    @Operation(
        summary = "Crear cuenta",
        description = "Crea una nueva cuenta bancaria",
        responses = {
            @ApiResponse(responseCode = "201", description = "Cuenta creada exitosamente"),
            @ApiResponse(responseCode = "400", description = "No se pudo crear la cuenta")
        }
    )
    public ResponseEntity<Object> create(@RequestBody CreateAccountDto dto) {
        return accountService.create(dto)
                .<ResponseEntity<Object>>map(account -> ResponseEntity.created(URI.create("/api/account/" + account.getId())).body(account))
                .orElse(ResponseEntity.badRequest().body("No se pudo crear la cuenta. Verifique que el usuario exista."));
    }

    @PutMapping("/{id}")
    @Operation(
        summary = "Actualizar cuenta",
        description = "Actualiza una cuenta bancaria existente",
        responses = {
            @ApiResponse(responseCode = "200", description = "Cuenta actualizada exitosamente"),
            @ApiResponse(responseCode = "404", description = "Cuenta no encontrada")
        }
    )
    public ResponseEntity<Object> update(@PathVariable Long id, @RequestBody UpdateAccountDto dto) {
        return accountService.update(id, dto)
                .<ResponseEntity<Object>>map(ResponseEntity::ok)
                .orElse(ResponseEntity.status(404).body("Cuenta no encontrada"));
    }

    @DeleteMapping("/{id}")
    @Operation(
        summary = "Eliminar cuenta",
        description = "Elimina una cuenta bancaria",
        responses = {
            @ApiResponse(responseCode = "204", description = "Cuenta eliminada exitosamente"),
            @ApiResponse(responseCode = "404", description = "Cuenta no encontrada")
        }
    )
    public ResponseEntity<Object> delete(@PathVariable Long id) {
        if (accountService.delete(id)) {
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.status(404).body("Cuenta no encontrada");
    }
}

File: Talleres RESFTUL JAVA DOTNET/API_GOOGLE_MAPS_REST_DOTNET_GR-03/04.SERVIDOR/API-DOTNET-BANK/Controllers/AccountController.cs
using API_DOTNET_BANK.Application.DTOs.Account;
using API_DOTNET_BANK.Application.Interface;
using API_DOTNET_BANK.Application.Service;
using API_DOTNET_BANK.Infraestructure;
using API_DOTNET_BANK.Infraestructure.Repository;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace API_DOTNET_BANK.Web.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class AccountController : ControllerBase
    {
        private readonly AppDbContext _dbContext;

        public AccountController(AppDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        private IAccountService CreateService()
        {
            var accountRepository = new AccountRepository(_dbContext);
            var userRepository = new UserRepository(_dbContext);
            return new AccountService(accountRepository, userRepository);
        }

        [HttpGet]
        public async Task<IActionResult> GetAll()
        {
            var accounts = await CreateService().GetAllAsync();
            return Ok(accounts);
        }

        [HttpGet("{id}")]
        public async Task<IActionResult> GetById(int id)
        {
            var account = await CreateService().GetByIdAsync(id);
            if (account == null)
                return NotFound("Cuenta no encontrada");
            return Ok(account);
        }

        [HttpGet("user/{userId}")]
        public async Task<IActionResult> GetByUserId(int userId)
        {
            var accounts = await CreateService().GetByUserIdAsync(userId);
            return Ok(accounts);
        }

        [HttpPost]
        public async Task<IActionResult> Create([FromBody] CreateAccountDto dto)
        {
            var account = await CreateService().CreateAsync(dto);
            if (account == null)
                return BadRequest("No se pudo crear la cuenta. Verifique que el usuario exista.");
            return CreatedAtAction(nameof(GetById), new { id = account.Id }, account);
        }

        [HttpPut("{id}")]
        public async Task<IActionResult> Update(int id, [FromBody] UpdateAccountDto dto)
        {
            var account = await CreateService().UpdateAsync(id, dto);
            if (account == null)
                return NotFound("Cuenta no encontrada");
            return Ok(account);
        }

        [HttpDelete("{id}")]
        public async Task<IActionResult> Delete(int id)
        {
            var result = await CreateService().DeleteAsync(id);
            if (!result)
                return NotFound("Cuenta no encontrada");
            return NoContent();
        }
    }
}
